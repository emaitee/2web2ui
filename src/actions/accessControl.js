import { fetch as fetchAccount } from './account';
import { get as getCurrentUser, getGrants } from './currentUser';

// initialize some state used for access control
export function initializeAccessControl() {
  return (dispatch) => Promise.all([
    dispatch(fetchAccount()),
    dispatch(getCurrentUser())
      .then(({ access_level }) => dispatch(getGrants({ role: access_level })))
  ])
    .then(() => dispatch({ type: 'ACCESS_CONTROL_READY' }))
    .catch((err) => {
    // TODO: log to SENTRY
    // for now, just ignore this error
    });
}
