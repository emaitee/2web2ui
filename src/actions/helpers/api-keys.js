import _ from 'lodash';
import { getGrants } from 'src/selectors/api-keys';

export function formatKeyForRequest(key, getState) {
  const request = { data: {}};

  if (key.subaccount) {
    request.headers = { 'X-MSYS-SUBACCOUNT': key.subaccount.id };
  }

  request.data.label = key.label;

  if (key.grantsRadio === 'all') {
    request.data.grants = _.keys(getGrants(getState()));
  } else {
    request.data.grants = _.keys(_.pickBy(key.grants));
  }

  if (key.validIps) {
    request.data.valid_ips = key.validIps.split(',').map(_.trim);
  }

  return request;
}
