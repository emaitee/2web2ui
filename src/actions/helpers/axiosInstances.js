import axios from 'axios';
import config from 'src/config';

const { apiBase, apiRequestTimeout, zuora: zuoraConfig } = config;

export const sparkpost = axios.create({
  baseURL: apiBase,
  timeout: apiRequestTimeout
});

export const zuora = axios.create({
  baseURL: zuoraConfig.baseUrl,
  timeout: zuoraConfig.timeout
});
