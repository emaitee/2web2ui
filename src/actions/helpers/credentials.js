import { pickBy, trim, keys } from 'lodash';

import { getGrants } from 'selectors/credentials';

export function formatKeyForRequest(key, getState) {
  const request = { data: {}};

  if (key.subaccount) {
    request.headers = { 'X-MSYS-SUBACCOUNT': key.subaccount.id };
  }

  request.data.label = key.label;

  if (key.grantsRadio === 'all') {
    request.data.grants = keys(getGrants(getState()));
  } else {
    request.data.grants = keys(pickBy(key.grants));
  }

  if (key.validIps) {
    request.data.valid_ips = key.validIps.split(',').map(trim);
  }

  return request;
}
