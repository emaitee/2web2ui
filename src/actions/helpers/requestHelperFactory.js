import axios from 'axios';

function defaultOnSuccess({ types, response, dispatch, meta, action }) {
  dispatch({
    type: types.SUCCESS,
    payload: response,
    meta
  });

  return response;
}

function defaultOnFail({ types, err, dispatch, meta, action }) {
  const { message, response = {}} = err;

  dispatch({
    type: types.FAIL,
    payload: { message, response },
    meta
  });
}

export default function requestFactory({
  request = axios,
  onSuccess = defaultOnSuccess,
  onFail = defaultOnFail,
  transformHttpOptions = (opts) => opts
} = {}) {
  return (action) => (dispatch, getState) => {
    const { type = 'NO_TYPE_DEFINED', meta } = action;
    const { url, method = 'get', params, headers, data } = meta;
    const types = {
      PENDING: `${type}_PENDING`,
      SUCCESS: `${type}_SUCCESS`,
      FAIL: `${type}_FAIL`
    };

    // standardize method included in meta from here on out
    meta.method = method.toLowerCase();

    dispatch({
      type: types.PENDING,
      meta
    });

    const httpOptions = {
      method: meta.method,
      url,
      params,
      headers,
      data
    };

    const transformedHttpOptions = transformHttpOptions(httpOptions, getState);

    return request(transformedHttpOptions)
      .then(
        // request succeeded, we only get here if the request returned a 2xx status code
        (response) => onSuccess({ types, response, dispatch, meta, action, getState }),

        // request failed (remember to throw err in your onFail)
        (err) => {
          onFail({ types, err, dispatch, meta, action, getState });
          return { error: err }
        }
      );
  };
}
