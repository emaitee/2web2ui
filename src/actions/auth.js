import { sparkpostLogin } from '../helpers/http';
import authCookie from '../helpers/authCookie';
import { initializeAccessControl } from './accessControl';


export function login(authData) {
  return (dispatch) => {
    dispatch({
      type: 'LOGIN_SUCCESS',
      payload: authData
    });

    dispatch(initializeAccessControl());
  };
}

export function authenticate(username, password, rememberMe = false) {
  // return a thunk
  return (dispatch, getState) => {
    const { loggedIn } = getState().auth;

    if (loggedIn) {
      return;
    }

    dispatch({ type: 'LOGIN_PENDING' });

    sparkpostLogin(username, password, rememberMe)
      .then(({ data = {}} = {}) => {
        const payload = { ...data, username };

        // save auth cookie
        authCookie.save(payload);

        // dispatch login success event
        dispatch(login(payload));
      })
      .catch((err) => {
        const { response = {}} = err;
        const { data = {}} = response;
        const { error_description: errorDescription } = data;

        // TODO: handle a timeout error better

        dispatch({
          type: 'LOGIN_FAIL',
          payload: {
            errorDescription
          }
        });
      });
  };
}

export function confirmPassword(username, password) {
  return (dispatch, getState) => {
    dispatch({ type: 'CONFIRM_PASSWORD' });

    return sparkpostLogin(username, password, false)
      .then(({ data = {}} = {}) => {
        const payload = { ...data, username };

        // dispatch login success event
        dispatch({
          type: 'CONFIRM_PASSWORD_SUCCESS',
          payload
        });
      })
      .catch((err) => {
        const { response = {}} = err;
        const { data = {}} = response;
        const { error_description: errorDescription } = data;
        dispatch({
          type: 'CONFIRM_PASSWORD_FAIL',
          payload: {
            errorDescription
          }
        });
        throw err;
      });
  };
}

export function beginRefresh(refreshToken) {
  return { type: 'ATTEMPTING_REFRESH', payload: refreshToken };
}

export function refresh(access_token, refresh_token) {
  const newCookie = authCookie.merge({ access_token, refresh_token });
  return login(newCookie);
}

export function logout() {
  return (dispatch) => {
    authCookie.remove();
    dispatch({
      type: 'LOGOUT'
    });
  };
}
