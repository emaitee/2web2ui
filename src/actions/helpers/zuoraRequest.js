import requestHelperFactory from './requestHelperFactory';
import _ from 'lodash';
import { zuora as zuoraAxios } from './axiosInstances';

export default requestHelperFactory({
  request: zuoraAxios,
  onSuccess: ({ types, response, dispatch, meta }) => {
    if (!response.data.success) {
      const message = _.get(response, 'data.reasons[0].message', 'An error occurred while contacting the billing service');

      dispatch({
        type: types.FAIL,
        payload: { message, response },
        meta
      });

      const error = new Error(message);
      error.response = response;

      return { error };
    }

    dispatch({
      type: types.SUCCESS,
      payload: response,
      meta
    });

    return response;
  }
});
