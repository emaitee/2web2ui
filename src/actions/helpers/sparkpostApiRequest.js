import requestHelperFactory from 'src/actions/helpers/requestHelperFactory';
import { refresh, logout } from 'src/actions/auth';
import { showAlert } from 'src/actions/globalAlert';
import { useRefreshToken } from 'src/helpers/http';
import { resolveOnCondition } from 'src/helpers/promise';
import _ from 'lodash';
import { sparkpost as sparkpostAxios } from './axiosInstances';

const maxRefreshRetries = 3;
export const refreshTokensUsed = new Set();
export let refreshing = false;

// Re-dispatches a given action after we finish refreshing the auth token
function redispatchAfterRefresh(action, dispatch) {
  return resolveOnCondition(() => !refreshing).then(() => dispatch(sparkpostRequest(action)));
}

const sparkpostRequest = requestHelperFactory({
  request: sparkpostAxios,
  transformHttpOptions: (options, getState) => {
    const { auth } = getState();
    const transformed = { ...options };
    if (auth.loggedIn) {
      _.set(transformed, 'headers.Authorization', auth.token);
    }
    return transformed;
  },
  onSuccess: ({ response, dispatch, types, meta }) => {
    const { data: { results }} = response;
    dispatch({
      type: types.SUCCESS,
      payload: results,
      meta
    });

    return results;
  },
  onFail: ({ types, err, dispatch, meta, action, getState }) => {
    const { message, response = {}} = err;
    const { auth } = getState();
    const { retries = 0 } = meta;

    // NOTE: if this is a 401 and we have a refresh token, we need to do a
    // refresh to get a new auth token and then re-dispatch this action
    if (response.status === 401 && auth.refreshToken && retries < maxRefreshRetries) {
      action.meta.retries = retries + 1;

      // If we are currently refreshing the token OR if this refresh token
      // has already been used to refresh, we should re-dispatch after refresh is complete
      if (refreshing || refreshTokensUsed.has(auth.refreshToken)) {
        return redispatchAfterRefresh(action, dispatch);
      }

      refreshing = true;
      refreshTokensUsed.add(auth.refreshToken);

      // call API for a new token
      return useRefreshToken(auth.refreshToken)

        // dispatch a refresh action to save new token results in cookie and store
        .then(({ data } = {}) => dispatch(refresh(data.access_token, data.refresh_token)))

        // dispatch the original action again, now that we have a new token ...
        // if anything in this refresh flow blew up, log out
        .then(
          // refresh token request succeeded
          () => {
            refreshing = false;
            return dispatch(sparkpostRequest(action));
          },
          // refresh token request failed
          (err) => {
            dispatch(logout());
            throw err;
          }
        );
    }

    // If we have a 403 or a 401 and we're not refreshing, log the user out silently
    if (response.status === 401 || response.status === 403) {
      dispatch(logout());
      throw err;
    }

    // any other API error should automatically fail, to be handled in the reducers/components
    dispatch({
      type: types.FAIL,
      payload: { message, response },
      meta
    });

    if (response.status >= 500) {
      dispatch(showAlert({ type: 'error', message: 'Something went wrong.', details: message }));
    }

    throw err;
  }
});

export default (action) => (dispatch, getState) => {
  // check for refreshing and exit early, otherwise call factory-made function
  if (refreshing) {
    return redispatchAfterRefresh(action, dispatch);
  }

  return dispatch(sparkpostRequest(action));
};
