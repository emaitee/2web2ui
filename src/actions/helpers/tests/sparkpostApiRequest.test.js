import sparkpostApiRequest, { refreshTokensUsed } from '../sparkpostApiRequest';
import { createMockStore } from 'src/__testHelpers__/mockStore';
import * as axiosMocks from '../axiosInstances';
import * as authMock from 'src/actions/auth';
import * as globalAlertMock from 'src/actions/globalAlert';
import * as httpHelpersMock from 'src/helpers/http';

jest.mock('../axiosInstances');
jest.mock('src/actions/auth');
jest.mock('src/actions/globalAlert');
jest.mock('src/helpers/http');

describe('Helper: SparkPost API Request', () => {

  let meta;
  let dispatchMock;
  let getStateMock;
  let results;
  let state;
  let action;
  let thunk;
  let refreshData;

  let mockStore;

  expect.hasAssertions();

  beforeEach(() => {
    jest.clearAllMocks();

    state = { auth: { loggedIn: true, token: 'TEST-TOKEN' }};
    mockStore = createMockStore(state);

    meta = { url: '/some/path', method: 'GET' };
    action = { type: 'TEST', meta };
    results = [];
    authMock.logout = jest.fn(() => ({ type: 'LOGOUT' }));

    axiosMocks.sparkpost.mockImplementation(() => Promise.resolve({ data: { results }}));
  });

  it('should successfully call the API', async () => {
    const actualResults = await mockStore.dispatch(sparkpostApiRequest(action));
    expect(actualResults).toBe(results);
    expect(mockStore.getActions()).toMatchSnapshot();
    expect(axiosMocks.sparkpost).toHaveBeenCalledWith(expect.objectContaining({
      method: 'get',
      url: '/some/path',
      headers: { Authorization: 'TEST-TOKEN' }
    }));
  });

  describe('failure cases', () => {

    let apiErr;

    beforeEach(() => {
      apiErr = new Error('API call failed');
      apiErr.response = { status: 400, data: { results }};
      axiosMocks.sparkpost.mockImplementation(() => Promise.reject(apiErr));
      refreshTokensUsed.clear();
    });

    it('should handle a failed API call', async () => {
      try {
        await mockStore.dispatch(sparkpostApiRequest(action));
      } catch (err) {
        expect(err).toBe(apiErr);
        expect(mockStore.getActions()).toMatchSnapshot();
      }
    });

    it('should dispatch a special 5xx error action', async () => {
      jest.spyOn(globalAlertMock, 'showAlert');
      apiErr.response.status = 500;
      try {
        await mockStore.dispatch(sparkpostApiRequest(action));
      } catch (err) {
        // const { message, response } = err;
        // expect(err).toBe(apiErr);
        expect(globalAlertMock.showAlert).toHaveBeenCalledWith({ message: 'Something went wrong.', type: 'error', details: apiErr.message});
        expect(mockStore.getActions()).toMatchSnapshot();
      }
      globalAlertMock.showAlert.mockRestore();
    });

    it('should dispatch a logout action on a 403 response', async () => {
      apiErr.response.status = 403;
      try {
        await mockStore.dispatch(sparkpostApiRequest(action));
      } catch (err) {
        expect(authMock.logout).toHaveBeenCalledTimes(1);
        expect(mockStore.getActions()).toMatchSnapshot();
      }
    });

    it('should dispatch a logout action on a 401 with no refresh token', async () => {
      apiErr.response.status = 401;
      try {
        await mockStore.dispatch(sparkpostApiRequest(action));
      } catch (err) {
        expect(authMock.logout).toHaveBeenCalledTimes(1);
        expect(mockStore.getActions()).toMatchSnapshot();
      }
    });

    describe('with refresh tokens', () => {

      beforeEach(() => {
        state.auth.refreshToken = 'REFRESH_1';
        mockStore = createMockStore(state);
        apiErr.response.status = 401;
        refreshData = { access_token: 'NEW_TOKEN', refresh_token: 'REFRESH_2' };
        httpHelpersMock.useRefreshToken = jest.fn(() => Promise.resolve({ data: refreshData }));
        authMock.refresh = jest.fn(() => ({ type: 'REFRESH' }));
      });

      it('should get a refresh token and re-dispatch', async () => {
        axiosMocks.sparkpost
          .mockImplementationOnce(() => Promise.reject(apiErr))
          .mockImplementation(() => Promise.resolve({ data: { results }}));

        await mockStore.dispatch(sparkpostApiRequest(action));
        expect(httpHelpersMock.useRefreshToken).toHaveBeenCalledTimes(1);
        expect(httpHelpersMock.useRefreshToken).toHaveBeenCalledWith('REFRESH_1');
        expect(authMock.refresh).toHaveBeenCalledWith('NEW_TOKEN', 'REFRESH_2');
        expect(action.meta.retries).toEqual(1);

        expect(mockStore.getActions()).toMatchSnapshot();
      });

      it('should only request one refresh token at a time', async () => {
        axiosMocks.sparkpost
          .mockImplementationOnce(() => Promise.reject(apiErr))
          .mockImplementationOnce(() => Promise.reject(apiErr))
          .mockImplementationOnce(() => Promise.reject(apiErr))
          .mockImplementation(() => Promise.resolve({ data: { results }}));

        await Promise.all([
          mockStore.dispatch(sparkpostApiRequest({ type: 'TEST_WITH_REFRESH', meta: {} })),
          mockStore.dispatch(sparkpostApiRequest({ type: 'TEST_NO_REFRESH', meta: {} })),
          mockStore.dispatch(sparkpostApiRequest({ type: 'TEST_NO_REFRESH', meta: {} }))
        ]);

        expect(httpHelpersMock.useRefreshToken).toHaveBeenCalledTimes(1);
        expect(httpHelpersMock.useRefreshToken).toHaveBeenCalledWith('REFRESH_1');
        expect(authMock.refresh).toHaveBeenCalledWith('NEW_TOKEN', 'REFRESH_2');

        // checking action counts instead of snapshotting them in order because these
        // are "kind of async" so they can sometimes get out of order and fail the snapshot
        const actions = mockStore.getActions();
        const byType = (keep) => ({ type }) => type === keep;
        expect(actions.filter(byType('TEST_WITH_REFRESH_PENDING'))).toHaveLength(2);
        expect(actions.filter(byType('TEST_WITH_REFRESH_SUCCESS'))).toHaveLength(1);
        expect(actions.filter(byType('REFRESH'))).toHaveLength(1);
        expect(actions.filter(byType('TEST_NO_REFRESH_PENDING'))).toHaveLength(4);
        expect(actions.filter(byType('TEST_NO_REFRESH_SUCCESS'))).toHaveLength(2);
      });

      it('should only retry 3 times', async () => {
        authMock.refresh.mockImplementation(() => {
          refreshTokensUsed.clear(); // have to do this since state always returns original refresh token
          return { type: 'REFRESH' };
        });

        try {
          await mockStore.dispatch(sparkpostApiRequest({ type: 'TEST_MAX_RETRIES', meta: {} }));
        } catch (err) {
          expect(httpHelpersMock.useRefreshToken).toHaveBeenCalledTimes(3);
          expect(httpHelpersMock.useRefreshToken).toHaveBeenCalledWith('REFRESH_1');
          expect(authMock.refresh).toHaveBeenCalledWith('NEW_TOKEN', 'REFRESH_2');
          expect(mockStore.getActions()).toMatchSnapshot();
        }
      });

      it('should only retry once with the same token', async () => {
        try {
          await mockStore.dispatch(sparkpostApiRequest({ type: 'TEST_ONE_RETRY_PER_TOKEN', meta: {} }));
        } catch (err) {
          expect(httpHelpersMock.useRefreshToken).toHaveBeenCalledTimes(1);
          expect(mockStore.getActions()).toMatchSnapshot();
        }
      });

      it('should log out if refresh fails', async () => {
        const refreshErr = new Error();
        httpHelpersMock.useRefreshToken = jest.fn(() => Promise.reject(refreshErr));

        try {
          await mockStore.dispatch(sparkpostApiRequest({ type: 'TEST_REFRESH_FAILED', meta: {}}));
        } catch (err) {
          expect(err).toBe(refreshErr);
          expect(authMock.logout).toHaveBeenCalledTimes(1);
        }
      });

    });

  });
});
