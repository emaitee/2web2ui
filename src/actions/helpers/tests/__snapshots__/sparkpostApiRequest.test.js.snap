// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper: SparkPost API Request failure cases should dispatch a logout action on a 401 with no refresh token 1`] = `
Array [
  Object {
    "meta": Object {
      "method": "GET",
      "url": "/some/path",
    },
    "type": "TEST_PENDING",
  },
  Object {
    "type": "LOGOUT",
  },
]
`;

exports[`Helper: SparkPost API Request failure cases should dispatch a logout action on a 403 response 1`] = `
Array [
  Object {
    "meta": Object {
      "method": "GET",
      "url": "/some/path",
    },
    "type": "TEST_PENDING",
  },
  Object {
    "type": "LOGOUT",
  },
]
`;

exports[`Helper: SparkPost API Request failure cases should dispatch a special 5xx error action 1`] = `
Array [
  Object {
    "meta": Object {
      "method": "GET",
      "url": "/some/path",
    },
    "type": "TEST_PENDING",
  },
  Object {
    "meta": Object {
      "method": "GET",
      "url": "/some/path",
    },
    "payload": Object {
      "message": "API call failed",
      "response": Object {
        "data": Object {
          "results": Array [],
        },
        "status": 500,
      },
    },
    "type": "TEST_FAIL",
  },
]
`;

exports[`Helper: SparkPost API Request failure cases should handle a failed API call 1`] = `
Array [
  Object {
    "meta": Object {
      "method": "GET",
      "url": "/some/path",
    },
    "type": "TEST_PENDING",
  },
  Object {
    "meta": Object {
      "method": "GET",
      "url": "/some/path",
    },
    "payload": Object {
      "message": "API call failed",
      "response": Object {
        "data": Object {
          "results": Array [],
        },
        "status": 400,
      },
    },
    "type": "TEST_FAIL",
  },
]
`;

exports[`Helper: SparkPost API Request failure cases with refresh tokens should get a refresh token and re-dispatch 1`] = `
Array [
  Object {
    "meta": Object {
      "method": "GET",
      "retries": 1,
      "url": "/some/path",
    },
    "type": "TEST_PENDING",
  },
  Object {
    "type": "REFRESH",
  },
  Object {
    "meta": Object {
      "method": "GET",
      "retries": 1,
      "url": "/some/path",
    },
    "type": "TEST_PENDING",
  },
  Object {
    "meta": Object {
      "method": "GET",
      "retries": 1,
      "url": "/some/path",
    },
    "payload": Array [],
    "type": "TEST_SUCCESS",
  },
]
`;

exports[`Helper: SparkPost API Request failure cases with refresh tokens should only retry 3 times 1`] = `
Array [
  Object {
    "meta": Object {
      "retries": 3,
    },
    "type": "TEST_MAX_RETRIES_PENDING",
  },
  Object {
    "type": "REFRESH",
  },
  Object {
    "meta": Object {
      "retries": 3,
    },
    "type": "TEST_MAX_RETRIES_PENDING",
  },
  Object {
    "type": "REFRESH",
  },
  Object {
    "meta": Object {
      "retries": 3,
    },
    "type": "TEST_MAX_RETRIES_PENDING",
  },
  Object {
    "type": "REFRESH",
  },
  Object {
    "meta": Object {
      "retries": 3,
    },
    "type": "TEST_MAX_RETRIES_PENDING",
  },
  Object {
    "type": "LOGOUT",
  },
]
`;

exports[`Helper: SparkPost API Request failure cases with refresh tokens should only retry once with the same token 1`] = `
Array [
  Object {
    "meta": Object {
      "retries": 3,
    },
    "type": "TEST_ONE_RETRY_PER_TOKEN_PENDING",
  },
  Object {
    "type": "REFRESH",
  },
  Object {
    "meta": Object {
      "retries": 3,
    },
    "type": "TEST_ONE_RETRY_PER_TOKEN_PENDING",
  },
  Object {
    "meta": Object {
      "retries": 3,
    },
    "type": "TEST_ONE_RETRY_PER_TOKEN_PENDING",
  },
  Object {
    "meta": Object {
      "retries": 3,
    },
    "type": "TEST_ONE_RETRY_PER_TOKEN_PENDING",
  },
  Object {
    "type": "LOGOUT",
  },
]
`;

exports[`Helper: SparkPost API Request should successfully call the API 1`] = `
Array [
  Object {
    "meta": Object {
      "method": "GET",
      "url": "/some/path",
    },
    "type": "TEST_PENDING",
  },
  Object {
    "meta": Object {
      "method": "GET",
      "url": "/some/path",
    },
    "payload": Array [],
    "type": "TEST_SUCCESS",
  },
]
`;
